### ISSUE TITLE
CS8601 false positive

### ISSUE THREAD

**Issue by MithrilMan** [dotnet/roslyn#41040]
#### Analyzer package

I suppose Microsoft.CodeQuality.CSharp.Analyzers

#### Package Version

v2.9.8 (Latest)

#### Diagnostic ID

CS8601	Possible null reference assignment.

#### Repro steps
```
private readonly Dictionary<int, string> values = new Dictionary<int, string>();
public bool TryGet(int index, [MaybeNullWhen(false)] out string value)
{
   return this.values.TryGetValue(index, out value);
}
```
#### Expected behavior
CS8601 shouldn't happen

#### Actual behavior
![image](https://user-images.githubusercontent.com/5107375/72632552-cc8d3b80-3956-11ea-9eca-84b79e78a913.png)

**Comment by sharwell** [comment]
Related to #38191. 

The expected signature of `TryGet` in the example is:

```csharp
public bool TryGet(int index, [NotNullWhen(false)] out string? value)
```

**Comment by MithrilMan** [comment]
where can I find documentation about implication between chosing
`[MaybeNullWhen(false)] out string value`
over
`[NotNullWhen(false)] out string? value`

to me it should have the same result in terms of CA, isnt't it?

**Comment by RikkiGibson** [comment]
It feels like this is related to #36039. My hope is that the behavior you are seeing is a rough edge that we will improve on.

/cc @jcouv who may wish to correct me or elaborate.

**Comment by MithrilMan** [comment]
Ok so actually you don't check method bodies for this rule and so since I assign null before returning, CA can't know if the value I'm setting to null is in a true or false path, correct?
So actually the only fix is to place the `!` like
`return this.values.TryGetValue(index, out value!);`

correct?
I'd like to get rid of unneeded `!` at some point :)

**Comment by jcouv** [comment]
Currently, nullability attributes affect callers, but are ignored while analyzing method bodies.

For instance `[MaybeNull] string M() { return null; }` will warn. Similarly, assigning `null` to an `[MaybeNullWhen(...)] out string parameter` will warn.

This is tracked by https://github.com/dotnet/roslyn/issues/36039 and will be addressed in 16.5p3 (PR https://github.com/dotnet/roslyn/pull/40789). For conditional attributes, we will not be able to analyze them "smartly", so we'll treat `[MaybeNullWhen(...)]` the same as `[MaybeNull]`. This will at least remove annoying warnings, although at a slight cost in safety.

After that, I think `[MaybeNullWhen(false)] out string value` and `[NotNullWhen(false)] out string? value` will be equivalent (both for consumers and implementers of the API).

I'll go ahead and close OP as duplicate of https://github.com/dotnet/roslyn/issues/36039 . Thanks

**Comment by MithrilMan** [comment]
thanks
